### ISSUE TITLE
Che-gwt migration to che-theia

### ISSUE THREAD

**Issue by fErten** [eclipse-che/che#13689]
### Description
Theia replaces the old GWT based IDE in Eclipse Che Version 7. What is the suggested flow for migrating custom extensions for previous Che versions (che 6.x) to new one?

These extensions based on che-6 are mostly [workspace-agent](https://github.com/eclipse/che/blob/master/assembly/assembly-wsagent-war/pom.xml) extension that is added to the assembly during the build.

### Reproduction Steps for Che-6
1- create custom-server-plugin (and extension to wsagent/http rest api written in java)
2- create custom-ide-plugin (gwt extension written in java)
3- add them to assembly-ide-war and assembly-wsagent-war during the build so that they are available in the customized ide.

Basically, the concept of the plugin that is defined [here](https://github.com/eclipse/che/tree/master/plugins).

### Migration
What I want to do is migrate the front-end and back-end functionality to Theia based IDE. 

For front-end such out-of-the-box migration is not possible, I need to add extend Theia-UI. My concern is here how can I reuse the server plugin with theia? Should expose it as a single service in side container? How theia can listen a new custom port for rest request? Does it has to be rest service? How this server plugin can be binded to the application?

Nevertheless, how to migrate che-6 server plugins to che 7 architecture?

**Comment by tsmaeder** [comment]
@fErten what kind of "server plugin" are we talking about here? I suspect you'll want your plugin to talk to some UI that is running inside the IDE? Do you need to access the service from somewhere else? In short, what are you trying to achieve?

**Comment by fErten** [comment]
@tsmaeder It is exactly how you described it actually, let me paint you the picture,

## Example Task
Access and list any remote machine file structure in a tree view through an ssh connection,

### How this is achieved in che-6

+ add remote-server-plugin
  - create a REST API (written in java)

 ```
POST /connection
DELETE /connection
GET /connection
GET /connection/{id}/files
UPDATE /connection
```

 - bind it to wsagent-server with guice library


`service is available under URL ${che-host}:${port}/api/connection
`

+ remote-ide-plugin
  - get wsagent endpoint and sent a request with related dto 
  - extend the gwt ide accordingly (tree widget, view, editor, etc)
  - parse retrieved response in order to display

+ make it part of the [assembly ](https://github.com/eclipse/che/tree/master/assembly)during the maven build

What I want is basically to reuse the rest api I've written.

I see that theia offers a framework for front-end with [TreeDataProvider ](https://github.com/eclipse/che-theia-samples/blob/master/samples/tree-view-sample-plugin/src/ftp-explorer.ts)to the parse the data, but with new structure I am confused for back-end. How to expose this service for theia, how to bind it? How theia-ui will listen to that service, how do I make the request?

Thanks for the help!

**Comment by fErten** [comment]
@tsmaeder Any thoughts regarding this matter?

**Comment by tsmaeder** [comment]
Hmhh...maybe this would work:

1. Make your server runnable standalone
2. Write a VS Code Plugin that can talk to your server on localhost. You'll need to familiarize yourself with https://code.visualstudio.com/api/references/contribution-points to determine whether the API is sufficient to do what you need
3. Package your server and it's prerequisites in an image based on the theia remote plugin endpoint base image
4. Make a Che 7 plugin based on that image including the VS Code plugin from 2)

**Comment by fErten** [comment]
@tsmaeder I see, thanks for the feedback. It seems like there is only a workaround to achieve that. Is it in your future plan to expose such API to be able to use the same backend-services in GWT and Theia IDE out-of-box or would it be chasing the rainbow?

**Comment by tsmaeder** [comment]
I am not aware that any feature work is planned for the GWT IDE.
